File: python/realtime/v2_compat.py
"""OpenAI Realtime GA / v2 leniency layer.

The Realtime API moved from beta to GA on 2025-08-28 and the beta namespace
was removed on 2026-05-07. The wire shape changed in three places that
affect us:

  • `session.type` is now a required discriminator (`"realtime"` vs
    `"transcription"`) -- beta sessions did not carry it.
  • `audio.input.{format,transcription,turn_detection,...}` and
    `audio.output.{format,voice,...}` replace the flat top-level keys
    `input_audio_format`, `input_audio_transcription`, `turn_detection`,
    `output_audio_format`, `voice`.
  • `output_modalities` replaces `modalities`. Same enum (`"audio"`,
    `"text"`), same semantics.

We're lenient: accept either shape on input, emit both on output, so old
and new clients both work. Unknown v2-only event types (DTMF, MCP-tool,
output_audio_buffer.*, classifier safety events) are accepted as no-ops
instead of returning UNKNOWN_EVENT_TYPE -- beta clients may not know about
them but it's wrong to 400 just because a client is more current than us.
"""
from __future__ import annotations

from typing import Any

KNOWN_V2_NOOP_EVENTS: frozenset[str] = frozenset({
    "output_audio_buffer.clear",
    "output_audio_buffer.append",
    "input_audio_buffer.dtmf.received",
    "transcription_session.update",
    "response.cancel_audio",
})

def normalize_session_object(session_obj: dict[str, Any]) -> dict[str, Any]:
    """Flatten v2 nested shape into the flat beta shape used internally.

    Idempotent: if the object is already flat, returns it unchanged (after
    a shallow copy). Never raises -- invalid types are left for the
    downstream parser to reject with its proper error code.
    """
    if not isinstance(session_obj, dict):
        return session_obj
    out: dict[str, Any] = dict(session_obj)

    audio = out.get("audio")
    if isinstance(audio, dict):
        ai = audio.get("input")
        if isinstance(ai, dict):
            if "format" in ai and "input_audio_format" not in out:
                out["input_audio_format"] = ai["format"]
            if "transcription" in ai and "input_audio_transcription" not in out:
                out["input_audio_transcription"] = ai["transcription"]
            if "turn_detection" in ai and "turn_detection" not in out:
                out["turn_detection"] = ai["turn_detection"]
        ao = audio.get("output")
        if isinstance(ao, dict):
            if "format" in ao and "output_audio_format" not in out:
                out["output_audio_format"] = ao["format"]
            if "voice" in ao and "voice" not in out:
                out["voice"] = ao["voice"]

    if "output_modalities" in out and "modalities" not in out:
        out["modalities"] = out["output_modalities"]

    return out

def enrich_session_view(view: dict[str, Any]) -> dict[str, Any]:
    """Add the v2 nested keys alongside the flat beta keys for emission.

    Mutates and returns `view`. Always ensures `session.type` is present
    (defaults to `"realtime"`).
    """
    if not isinstance(view, dict):
        return view

    view.setdefault("type", "realtime")

    audio = view.get("audio")
    if not isinstance(audio, dict):
        audio = {}
        view["audio"] = audio

    inp = audio.get("input")
    if not isinstance(inp, dict):
        inp = {}
        audio["input"] = inp
    if "input_audio_format" in view:
        inp.setdefault("format", view["input_audio_format"])
    if "input_audio_transcription" in view:
        inp.setdefault("transcription", view["input_audio_transcription"])
    if "turn_detection" in view:
        inp.setdefault("turn_detection", view["turn_detection"])

    outp = audio.get("output")
    if not isinstance(outp, dict):
        outp = {}
        audio["output"] = outp
    if "output_audio_format" in view:
        outp.setdefault("format", view["output_audio_format"])
    if "voice" in view:
        outp.setdefault("voice", view["voice"])

    if "modalities" in view:
        view.setdefault("output_modalities", view["modalities"])

    return view

def is_known_v2_noop_event(event_type: str) -> bool:
    """True for v2 event types we accept but do not act on."""
    return event_type in KNOWN_V2_NOOP_EVENTS

File: rust/src/realtime/v2_compat.rs
use serde_json::{Map, Value};

const KNOWN_V2_NOOP_EVENTS: &[&str] = &[
    "output_audio_buffer.clear",
    "output_audio_buffer.append",
    "input_audio_buffer.dtmf.received",
    "transcription_session.update",
    "response.cancel_audio",
];

pub(super) fn normalize_session_object(session_obj: &Value) -> Value {
    let Some(obj) = session_obj.as_object() else {
        return session_obj.clone();
    };
    let mut out: Map<String, Value> = obj.clone();

    if let Some(audio) = obj.get("audio").and_then(|v| v.as_object()) {
        if let Some(ai) = audio.get("input").and_then(|v| v.as_object()) {
            if let Some(v) = ai.get("format") {
                out.entry("input_audio_format").or_insert_with(|| v.clone());
            }
            if let Some(v) = ai.get("transcription") {
                out.entry("input_audio_transcription")
                    .or_insert_with(|| v.clone());
            }
            if let Some(v) = ai.get("turn_detection") {
                out.entry("turn_detection").or_insert_with(|| v.clone());
            }
        }
        if let Some(ao) = audio.get("output").and_then(|v| v.as_object()) {
            if let Some(v) = ao.get("format") {
                out.entry("output_audio_format")
                    .or_insert_with(|| v.clone());
            }
            if let Some(v) = ao.get("voice") {
                out.entry("voice").or_insert_with(|| v.clone());
            }
        }
    }

    if obj.contains_key("output_modalities") && !obj.contains_key("modalities") {
        if let Some(v) = obj.get("output_modalities") {
            out.insert("modalities".into(), v.clone());
        }
    }

    Value::Object(out)
}

pub(super) fn enrich_session_view(view: &mut Value) {
    let Some(obj) = view.as_object_mut() else {
        return;
    };
    obj.entry("type")
        .or_insert_with(|| Value::String("realtime".into()));

    let input_audio_format = obj.get("input_audio_format").cloned();
    let input_audio_transcription = obj.get("input_audio_transcription").cloned();
    let turn_detection = obj.get("turn_detection").cloned();
    let output_audio_format = obj.get("output_audio_format").cloned();
    let voice = obj.get("voice").cloned();
    let modalities = obj.get("modalities").cloned();

    let audio_entry = obj
        .entry("audio")
        .or_insert_with(|| Value::Object(Map::new()));
    if let Some(audio_obj) = audio_entry.as_object_mut() {
        let inp = audio_obj
            .entry("input")
            .or_insert_with(|| Value::Object(Map::new()));
        if let Some(inp_obj) = inp.as_object_mut() {
            if let Some(v) = input_audio_format {
                inp_obj.entry("format").or_insert(v);
            }
            if let Some(v) = input_audio_transcription {
                inp_obj.entry("transcription").or_insert(v);
            }
            if let Some(v) = turn_detection {
                inp_obj.entry("turn_detection").or_insert(v);
            }
        }
        let outp = audio_obj
            .entry("output")
            .or_insert_with(|| Value::Object(Map::new()));
        if let Some(outp_obj) = outp.as_object_mut() {
            if let Some(v) = output_audio_format {
                outp_obj.entry("format").or_insert(v);
            }
            if let Some(v) = voice {
                outp_obj.entry("voice").or_insert(v);
            }
        }
    }

    if let Some(m) = modalities {
        obj.entry("output_modalities").or_insert(m);
    }
}

pub(super) fn is_known_v2_noop_event(event_type: &str) -> bool {
    KNOWN_V2_NOOP_EVENTS.contains(&event_type)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn normalize_nested_to_flat() {
        let nested = json!({
            "type": "realtime",
            "audio": {
                "input":  {"format": "pcm16", "turn_detection": {"type": "server_vad"}},
                "output": {"format": "g711_ulaw", "voice": "alloy"},
            },
            "output_modalities": ["audio"],
        });
        let flat = normalize_session_object(&nested);
        assert_eq!(flat["input_audio_format"], "pcm16");
        assert_eq!(flat["output_audio_format"], "g711_ulaw");
        assert_eq!(flat["voice"], "alloy");
        assert_eq!(flat["turn_detection"]["type"], "server_vad");
        assert_eq!(flat["modalities"], json!(["audio"]));
    }

    #[test]
    fn normalize_is_idempotent_on_flat() {
        let flat = json!({"input_audio_format": "pcm16"});
        let again = normalize_session_object(&flat);
        assert_eq!(again, flat);
    }

    #[test]
    fn enrich_adds_nested_shape() {
        let mut v = json!({
            "id": "sess_x",
            "input_audio_format": "pcm16",
            "output_audio_format": "pcm16",
            "voice": "alloy",
            "modalities": ["audio", "text"],
        });
        enrich_session_view(&mut v);
        assert_eq!(v["type"], "realtime");
        assert_eq!(v["audio"]["input"]["format"], "pcm16");
        assert_eq!(v["audio"]["output"]["format"], "pcm16");
        assert_eq!(v["audio"]["output"]["voice"], "alloy");
        assert_eq!(v["output_modalities"], json!(["audio", "text"]));
    }
}
